File: zero_sum_game/dom_zero_agent.py
from dom_m1_agent import *


class DoMZeroPlayer:
    def __init__(self, game_duration:int, softmax_temperature:float, discount_factor:float) -> None:
        self.game_duration = game_duration
        self.softmax_temperature = softmax_temperature
        self.discount_factor = discount_factor
        self.actions = []
        self.beliefs = []
        self.dom_level = 0

    def irl(self, prior: np.array, observation: int, iteration: int) -> np.array:
        p_0 = 1/2
        p_1 = softmax_transformation(dom_m1_utility(game_1), self.softmax_temperature)[observation]
        p_2 = softmax_transformation(dom_m1_utility(game_2), self.softmax_temperature)[observation]
        unnormalized_posterior = prior * np.array([p_0, p_1, p_2])
        normalized_posterior = unnormalized_posterior / np.sum(unnormalized_posterior)
        return normalized_posterior

    @staticmethod
    def dom_zero_utility_subroutine(action: int, policies_matrix: np.array):
        uninformed_payoff = 1/2 * np.matmul(game_1[:, action], policies_matrix[0, :]) + \
                            1/2 * np.matmul(game_2[:, action], policies_matrix[0, :])
        p_1_payoff = np.matmul(game_1[:, action], policies_matrix[1, :])
        p_2_payoff = np.matmul(game_2[:, action], policies_matrix[2, :])
        return uninformed_payoff, p_1_payoff, p_2_payoff

    def act(self, belief):
        # Optimal policies per type        
        pi_0 = softmax_transformation(np.array([1/2, 1/2]), self.softmax_temperature)
        pi_1 = softmax_transformation(dom_m1_utility(game_1), self.softmax_temperature)
        pi_2 = softmax_transformation(dom_m1_utility(game_2), self.softmax_temperature)
        policies_matrix = np.array([pi_0, pi_1, pi_2])
        u_1 = -1 * np.matmul(belief, self.dom_zero_utility_subroutine(0, policies_matrix))
        u_2 = -1 * np.matmul(belief, self.dom_zero_utility_subroutine(1, policies_matrix))
        u_3 = -1 * np.matmul(belief, self.dom_zero_utility_subroutine(2, policies_matrix))
        return softmax_transformation(np.array([u_1, u_2, u_3]), self.softmax_temperature)

File: zero_sum_game/dom_zero_agent.R
dom_0_irl <- function(prior, observation)
{
  p_0 = 1/2
  p_1 = softmax(dom_m1_utility(game_1))[observation]
  p_2 = softmax(dom_m1_utility(game_2))[observation]
  unnormalized_posterior = prior * c(p_0, p_1, p_2)
  normalized_posterior = unnormalized_posterior / sum(unnormalized_posterior) 
  return(normalized_posterior)
}

dom_o_utility_subroutine <- function(action, policies_matrix)
{
  random_payoff = 1/2 * game_1[,action] %*% policies_matrix[1,] + 1/2 * game_2[,action] %*% policies_matrix[1,]
  p_1_payoff = game_1[,action] %*% policies_matrix[2,]
  p_2_payoff = game_2[,action] %*% policies_matrix[3,]
  
  return(c(random_payoff,p_1_payoff,p_2_payoff))
}

dom_0_utility <- function(belief)
{
  # Optimal policies per type
  pi_0 = t(c(1/2, 1/2))
  pi_1 = softmax(dom_m1_utility(game_1))
  pi_2 = softmax(dom_m1_utility(game_2))
  
  policies_matrix <- matrix(c(pi_0, pi_1, pi_2), nrow = 3, byrow = T)
  
  u_1 = -1 * sum(belief %*% dom_o_utility_subroutine(1, policies_matrix))
  u_2 = -1 * sum(belief %*% dom_o_utility_subroutine(2, policies_matrix))
  u_3 = -1 * sum(belief %*% dom_o_utility_subroutine(3, policies_matrix))
  
  return(c(u_1, u_2, u_3))
}
